### PR TITLE
Avoid null-pointer arithmetic, amended.

### PR DESCRIPTION
**Pull request by rpoyner-tri**
The first attempt (PR#1) inadvertently made some implementation behavior
unreachable.

The second attempt (PR#2) failed debug-time assertion checks.

### CODE REVIEW

File: picosat.c [hunk 1]
```diff
@@ -2816,10 +2816,12 @@ fix_impl_lits (PS * ps, long delta)
...
-  for (s = ps->impls + 2; s <= ps->impls + 2 * ps->max_var + 1; s++)
-    if (s->start)
-      for (p = s->start; p < s->start + s->count; p++)
-        *p += delta;
+  for (s = ps->impls + 2; s <= ps->impls + 2 * ps->max_var + 1; s++) {
+    if (!s->start) { continue; }
+    for (p = s->start; p < s->start + s->count; p++) {
+      *p += delta;
+    }
+  }
```
Comment (mwoehlke-kitware) [inline picosat.c]: Besides adding braces, what does this change do?

File: picosat.c [hunk 2]
```diff
@@ -3430,9 +3432,9 @@ satisfied (PS * ps)
     return 0;
   assert (!ps->conflict);
   assert (bcp_queue_is_empty (ps));
-  if (ps->trail == NULL)
-    return ps->thead == NULL;
-  return ps->thead == ps->trail + ps->max_var;	/* all assigned */
+  return ps->trail
+      ? (ps->thead == ps->trail + ps->max_var) /* all assigned */
+      : (ps->max_var == 0); /* ps->trail was NULL, so avoid ptr arithmetic */
```
Comment (mwoehlke-kitware) [inline picosat.c:3435]: Neither of these are obviously correct. I take it it is (theoretically) guaranteed that `trail == NULL` ⇒ `thead == NULL`? Strictly speaking, the `:` check "should" be `(intptr_t)ps->thead == ps->max_var`, though logically of course that looks dodgy. But if the preceding implication holds, `ps->max_var == 0` is equivalent.

File: picosat.c [hunk 3]
```diff
@@ -3903,43 +3905,40 @@ prop2 (PS * ps, Lit * this)
 
 #ifdef NO_BINARY_CLAUSES
   lstk = LIT2IMPLS (this);
-  if (lstk->count)
+  start = lstk->start;
+  l = start ? start + lstk->count : start;
```
Comment (mwoehlke-kitware) [inline picosat.c]: What is the purpose of this change? It seems to make the logic flow less clear to no obvious benefit.

File: picosat.c [hunk 4]
```diff
@@ -4714,7 +4713,7 @@ collect_clauses (PS * ps)
 	      Ltk * lstk = LIT2IMPLS (lit);
 	      Lit ** r, ** s;
 	      r = lstk->start;
-	      if (lstk->count && (lit->val != TRUE || LIT2VAR (lit)->level))
+	      if (r && (lit->val != TRUE || LIT2VAR (lit)->level))
```
Comment (mwoehlke-kitware) [inline picosat.c]: I don't understand this change. If `r` is non-null, but `lstk->count == 0`, the loop will not execute anyway. If `r` is null, and `lstk->count != 0`, this code would SEGV, so I presume that doesn't happen. I'm not sure this change is beneficial?

File: picosat.c (hunk 1 above)
Comment (mwoehlke-kitware) [reply picosat.c]: Also, the braces here are malformatted.

Comment (rpoyner-tri) [reply picosat.c]: The `continue` on 2820 is the crucial change.

I will work on brace formatting.

Comment (mwoehlke-kitware) [reply picosat.c]: But what's the difference between `for (...) { if (!s->start) continue; do_stuff; }` and `for (...) { if (s->start) do_stuff; }`?

File: picosat.c (hunk 2 above)
Comment (rpoyner-tri) [reply picosat.c:3435]: So, we could just go `(intptr_t)` everywhere on the original expression; that might be more clear. However, I actually tried it.

```
return (intptr_t)ps->thead == (intptr_t)ps->trail + ps->max_var;
```

and got this from ubsan, elsewhere:

```
picosat.c:3920:15: runtime error: load of misaligned address 0x00035341a2f9 for type 'Lit *' (aka 'struct Lit *'), which requires 8 byte alignment                                                
```
so just slapping casts on things is clearly giving wrong answers.

My reasoning is that `ps->trail` and `ps->thead` are logically part of the same "growable array" data structure, where `ps->trail` owns the storage and `thead` and `eot` are non-owning pointers into the same storage at different offsets. I infer this from the `ENLARGE` invocation within `tpush()`, which is the only place where `trail` ever gets storage; and from the `DELETEN (ps->trail, /*...*/)` statement inside `reset()`.

FWIW, I don't really understand why/when `trail==NULL` and `ps->max_var!=0`, but an intermediate attempt (see PR#2) failed assertions within picosat.

File: picosat.c (hunk 3 above)
Comment (rpoyner-tri) [reply picosat.c]: It checks for null before the UB `start + lstk->count`, and no-ops the subsequent while loop.

File: picosat.c (hunk 4 above)
Comment (rpoyner-tri) [reply picosat.c]: `r` (and later `s`) are aliases for `lstk->start` or offsets into its owned storage. If lstk->start is NULL at 4716, then `r` will be NULL at the if check, so the condition will fail and the contained for loop be skipped.

Since lstk->count is clearly a "how many elements" metric on lstk->start, it is either 0 or garbage if lstk->start is NULL.

File: picosat.c (hunk 2 above)
Comment (mwoehlke-kitware) [reply picosat.c:3435]: > ...I actually tried it

Adding an offset to a pointer is not the same as adding that integer to the pointer converted to an integer. The offset gets multiplied by `sizeof(*p)`. That's likely why your other attempt broke. (Hmm, mine has the same error :flushed:.)

But that's something of a red herring. If you're convinced that `trail == NULL` ⇒ `thead == NULL`, this is probably fine. My point was more to verify that you *are* convinced that implication holds.

File: picosat.c (hunk 3 above)
Comment (mwoehlke-kitware) [reply picosat.c]: ...but if `lstk->count` is non-zero, the original code would SEGV, yes? Therefore, `!start` ⇒ `!count`. Since the old code cuts out the `while` loop, which is functionally equivalent to short-circuiting it, both versions appear to have the same effect?

Ultimately, it looks like a number of these changes result from an independent, pre-#1 attempt to fix things. But since #1 was merged, if two fixes are functionally equivalent, I think it would be simpler to keep the changes that are already merged, in order to reduce further churn and simplify review. (Applies here, at L4716, and L2819-2824.)

File: picosat.c (hunk 4 above)
Comment (mwoehlke-kitware) [reply picosat.c]: ...if it's garbage, the original could would SEGV, yes? As noted above, it seems that `!start` ⇒ `!count` and this change has no functional effect. If that impression is incorrect, I'd really like to understand *how*.

(Your correct-AFAICT observation that `count` "is either 0 or garbage" would seem to confirm the stated implication.)

File: picosat.c (hunk 1 above)
Comment (rpoyner-tri) [reply picosat.c]: You are right, there is no functional difference here. The continue formulation saves one level of indent (4 vs. 3).

I happened to start from a pre-PR#1 source file, and came to a different, but equivalent patch here. I can revert this one.

Comment (rpoyner-tri) [reply picosat.c]: Reverted my changes in favor of text from PR#1.

File: picosat.c (hunk 2 above)
Comment (rpoyner-tri) [reply picosat.c:3435]: I'm satisfied this change is correct, and that  `trail == NULL` ⇒ `thead == NULL`. They both start and lazily remain NULL (from init() {... memset() ...}), both get set by tpush() { ... ENLARGE ...} and then the only place `trail` gets cleared again is in  reset() { ... DELETEN ...}, but reset() also deletes the entire containing structure.

File: picosat.c (hunk 4 above)
Comment (rpoyner-tri) [reply picosat.c]: Reverted my changes in favor of text from PR#1.

File: picosat.c (hunk 3 above)
Comment (rpoyner-tri) [reply picosat.c]: Reverted my change in favor of text from PR#1.

**Review by mwoehlke-kitware — APPROVED** [review]
LGTM, thanks for simplifying!

**Comment by mwoehlke-kitware** [comment]
Thanks, @soonhokong!
